core/distributed/vector_kernels.hpp

```diff
@@ -0,0 +1,82 @@
+Copyright (c) 2017-2022, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_DISTRIBUTED_VECTOR_KERNELS_HPP_
+#define GKO_CORE_DISTRIBUTED_VECTOR_KERNELS_HPP_
+
+
+// can't include ginkgo/core/distributed/vector.hpp since that requires linking
+// against MPI
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/distributed/partition.hpp>
+
+
+#include "core/base/kernel_declaration.hpp"
+
+
+namespace gko {
+namespace kernels {
+
+
+#define GKO_DECLARE_DISTRIBUTED_VECTOR_BUILD_LOCAL(ValueType, LocalIndexType, \
+                                                   GlobalIndexType)           \
+    void build_local(                                                         \
+        std::shared_ptr<const DefaultExecutor> exec,                          \
+        const Array<matrix_data_entry<ValueType, GlobalIndexType>>& input,    \
+        const distributed::Partition<LocalIndexType, GlobalIndexType>*        \
+            partition,                                                        \
+        comm_index_type local_part,                                           \
+        Array<matrix_data_entry<ValueType, LocalIndexType>>& local_data,      \
+        ValueType deduction_help)
+
+
+#define GKO_DECLARE_ALL_AS_TEMPLATES                                      \
+    using comm_index_type = distributed::comm_index_type;                 \
```

Comment:
This is probably better defined in one of the headers in the `distributed` namespace

Reply:
This is mostly to shorten the kernel definition. I wouldn't put that in another header, since that shouldn't be publicly available. I could also just remove it.